# SI 618 Homework 1

## Introduction: (5 points)

### Q1: Mount drive and change directory to where ‘book_data.csv’ is located (0.5 point)

In [3]:
# 查看目前的工作路径
import os
path = os.getcwd()
print(path)

/Users/peilinqiu/Desktop/Python/SI618/Homework Assignments/Homework1/Homework1


In [4]:
# 更改当前工作路径
os.chdir('/Users/peilinqiu/Desktop/Python/SI618/Homework Assignments/Homework1/Homework1')
print(os.getcwd())

/Users/peilinqiu/Desktop/Python/SI618/Homework Assignments/Homework1/Homework1


### Q2: Load the dataset. (0.5 point)

In [5]:
import pandas as pd
import numpy as np

In [6]:
book_data=pd.read_csv("book_data.csv")

In [7]:
book_data.head()

book_authors                                          book_desc  \
0   Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1  Madeline Miller  In the house of Helios, god of the sun and mig...   
2      Holly Black  Of course I want to be like them. They’re beau...   
3        A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4    Tara Westover  An unforgettable memoir in the tradition of Th...   

  book_edition     book_format     book_isbn book_pages  book_rating  \
0          NaN  Kindle Edition           NaN  435 pages         4.33   
1          NaN       Hardcover  9.780000e+12  393 pages         4.34   
2          NaN       Hardcover  9.780000e+12  370 pages         4.18   
3          NaN  Kindle Edition           NaN  449 pages         3.97   
4          NaN       Hardcover  9.780000e+12  334 pages         4.47   

   book_rating_count  book_review_count               book_title  \
0             146505              17438          The Great Alone   
1              61357               9369                    Circe   
2              62897              11948         The Cruel Prince   
3             175678              20645  The Woman in the Window   
4             128337              16424                 Educated   

                                              genres  \
0  Fiction|Historical|Historical Fiction|Audioboo...   
1  Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                    Fantasy|Young Adult|Fairies|Fae   
3  Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4          Nonfiction|Autobiography|Memoir|Biography   

                                           image_url  
0  https://images.gr-assets.com/books/1501852423l...  
1  https://images.gr-assets.com/books/1508879575l...  
2  https://images.gr-assets.com/books/1493047153l...  
3  https://images.gr-assets.com/books/1528225499l...  
4  https://images.gr-assets.com/books/1506026635l...

### Q3: How many books are there? (2 points)

In [8]:
# 查询具体某一列的count
book_data[['book_title']].count()

book_title    1141
dtype: int64

In [9]:
# 查询所有列分别的count
book_data.count()

book_authors         1141
book_desc            1136
book_edition          105
book_format          1107
book_isbn             691
book_pages           1071
book_rating          1141
book_rating_count    1141
book_review_count    1141
book_title           1141
genres               1130
image_url            1137
dtype: int64

### Q4: How many books have more than 1 author? (2 points)

In [226]:
## 仇培林答案
author_num = book_data['book_authors'].str.split('|', expand=True).stack().value_counts()

In [227]:
author_num

K.  Webster                  15
Samantha Chase                6
Penny Reid                    5
Stephen King                  5
Vi Keeland                    5
                             ..
Mary H.K. Choi                1
Florence Gonsalves            1
Catherynne M. Valente         1
Wendy Heard                   1
Jennifer Zeynab Joukhadar     1
Length: 1212, dtype: int64

In [228]:
author_num.head(15)

K.  Webster        15
Samantha Chase      6
Penny Reid          5
Stephen King        5
Vi Keeland          5
Penelope Ward       5
Mike Mignola        5
Tijan               4
Jill Shalvis        4
N.R. Walker         4
M.A. Innes          4
James Patterson     4
Julie Mulhern       4
Stuart Woods        4
Karina Halle        4
dtype: int64

In [229]:
author_num[author_num > 1].count() 

162

In [10]:
## 杨丹亭答案
book_data['author_num'] = book_data['book_authors'].apply(lambda x: x.count('|')+1)
book_data.head(10)

book_authors  \
0                                 Kristin Hannah   
1                                Madeline Miller   
2                                    Holly Black   
3                                      A.J. Finn   
4                                  Tara Westover   
5                                   Tomi Adeyemi   
6                                   Tayari Jones   
7  Michelle McNamara|Gillian Flynn|Patton Oswalt   
8                 Greer Hendricks|Sarah Pekkanen   
9                                Neal Shusterman   

                                           book_desc book_edition  \
0  Alaska, 1974.Unpredictable. Unforgiving. Untam...          NaN   
1  In the house of Helios, god of the sun and mig...          NaN   
2  Of course I want to be like them. They’re beau...          NaN   
3  Anna Fox lives alone—a recluse in her New York...          NaN   
4  An unforgettable memoir in the tradition of Th...          NaN   
5  They killed my mother.They took our magic.They...          NaN   
6  Newlyweds Celestial and Roy are the embodiment...          NaN   
7  A masterful true crime account of the Golden S...          NaN   
8  When you read this book, you will make many as...          NaN   
9  Rowan has gone rogue, and has taken it upon hi...          NaN   

      book_format     book_isbn book_pages  book_rating  book_rating_count  \
0  Kindle Edition           NaN  435 pages         4.33             146505   
1       Hardcover  9.780000e+12  393 pages         4.34              61357   
2       Hardcover  9.780000e+12  370 pages         4.18              62897   
3  Kindle Edition           NaN  449 pages         3.97             175678   
4       Hardcover  9.780000e+12  334 pages         4.47             128337   
5       Hardcover  9.780000e+12  525 pages         4.25              56789   
6       Hardcover  9.780000e+12  308 pages         4.01              91515   
7           ebook  9.780000e+12  352 pages         4.22              60784   
8       Hardcover  9.780000e+12  346 pages         3.85              89835   
9       Hardcover  9.780000e+12  504 pages         4.53              22823   

   book_review_count                                         book_title  \
0              17438                                    The Great Alone   
1               9369                                              Circe   
2              11948                                   The Cruel Prince   
3              20645                            The Woman in the Window   
4              16424                                           Educated   
5              12046                         Children of Blood and Bone   
6               9836                               An American Marriage   
7               7148  I'll Be Gone in the Dark: One Woman's Obsessiv...   
8              11331                                The Wife Between Us   
9               4620                                        Thunderhead   

                                              genres  \
0  Fiction|Historical|Historical Fiction|Audioboo...   
1  Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                    Fantasy|Young Adult|Fairies|Fae   
3  Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4          Nonfiction|Autobiography|Memoir|Biography   
5                        Fantasy|Young Adult|Fiction   
6                     Fiction|Contemporary|Audiobook   
7  Nonfiction|Crime|True Crime|Mystery|Crime|Myst...   
8  Thriller|Fiction|Mystery|Suspense|Thriller|Mys...   
9  Young Adult|Science Fiction|Science Fiction|Dy...   

                                           image_url  author_num  
0  https://images.gr-assets.com/books/1501852423l...           1  
1  https://images.gr-assets.com/books/1508879575l...           1  
2  https://images.gr-assets.com/books/1493047153l...           1  
3  https://images.gr-assets.com/books/1528225499l...           1  
4  https://images.gr-assets.com/books/1506026635l...           1  

In [18]:
book_data.to_excel('book_data.xlsx', index = None, header=True)

In [15]:
num = book_data['author_num'] > 1
num.value_counts()

False    997
True     144
Name: author_num, dtype: int64

## Length: (30 points)

### Q5: You will want to create a new column with the integer value of the number of pages (15 points)

In [230]:
book_data.dtypes

book_authors          object
book_desc             object
book_edition          object
book_format           object
book_isbn            float64
book_pages            object
book_rating          float64
book_rating_count      int64
book_review_count      int64
book_title            object
genres                object
image_url             object
dtype: object

In [231]:
book_data['book_pages'].astype(str)

0       435 pages
1       393 pages
2       370 pages
3       449 pages
4       334 pages
          ...    
1136    304 pages
1137    417 pages
1138    320 pages
1139    419 pages
1140    304 pages
Name: book_pages, Length: 1141, dtype: object

In [232]:
book_data['book_pages'] = book_data.book_pages.str.replace(r"[a-zA-Z]",'')

In [233]:
book_data.head()

book_authors                                          book_desc  \
0   Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1  Madeline Miller  In the house of Helios, god of the sun and mig...   
2      Holly Black  Of course I want to be like them. They’re beau...   
3        A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4    Tara Westover  An unforgettable memoir in the tradition of Th...   

  book_edition     book_format     book_isbn book_pages  book_rating  \
0          NaN  Kindle Edition           NaN       435          4.33   
1          NaN       Hardcover  9.780000e+12       393          4.34   
2          NaN       Hardcover  9.780000e+12       370          4.18   
3          NaN  Kindle Edition           NaN       449          3.97   
4          NaN       Hardcover  9.780000e+12       334          4.47   

   book_rating_count  book_review_count               book_title  \
0             146505              17438          The Great Alone   
1              61357               9369                    Circe   
2              62897              11948         The Cruel Prince   
3             175678              20645  The Woman in the Window   
4             128337              16424                 Educated   

                                              genres  \
0  Fiction|Historical|Historical Fiction|Audioboo...   
1  Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                    Fantasy|Young Adult|Fairies|Fae   
3  Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4          Nonfiction|Autobiography|Memoir|Biography   

                                           image_url  
0  https://images.gr-assets.com/books/1501852423l...  
1  https://images.gr-assets.com/books/1508879575l...  
2  https://images.gr-assets.com/books/1493047153l...  
3  https://images.gr-assets.com/books/1528225499l...  
4  https://images.gr-assets.com/books/1506026635l...

In [234]:
book_data.dtypes

book_authors          object
book_desc             object
book_edition          object
book_format           object
book_isbn            float64
book_pages            object
book_rating          float64
book_rating_count      int64
book_review_count      int64
book_title            object
genres                object
image_url             object
dtype: object

In [235]:
print(book_data[book_data['book_pages'].isnull()])

            book_authors                                          book_desc  \
15       Shantel Tessier  My name is Avery Decker, and my story isn’t fo...   
20       Shantel Tessier  What do you do when the devil has you in his s...   
142          Ella Fields  Callum Welsh perfected the art of hating me wh...   
147         Harper Sloan  Ari Daniels didn’t count on her whole world tu...   
178         Meghan Quinn  "Good Morning Malibu, it’s another beautiful d...   
...                  ...                                                ...   
1019       C.L. Matthews  She’s forbidden.Too young, too pure, and not m...   
1021  Jonathan Kellerman  A disturbing murder with shocking consequences...   
1023          Ryan North  "How to Invent Everything is such a cool book....   
1097           Joy Ellis  The most difficult case of Detective Jackman’s...   
1113     Waubgeshig Rice  With winter looming, a small northern Anishina...   

     book_edition     book_format     book_isbn boo

In [236]:
book_data = book_data[book_data['book_pages'].notna()]

In [237]:
book_data

book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4        Tara Westover  An unforgettable memoir in the tradition of Th...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1139  Sarah Beth Durst  The battle between vicious spirits and strong-...   
1140      Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition     book_format     book_isbn book_pages  book_rating  \
0             NaN  Kindle Edition           NaN       435          4.33   
1             NaN       Hardcover  9.780000e+12       393          4.34   
2             NaN       Hardcover  9.780000e+12       370          4.18   
3             NaN  Kindle Edition           NaN       449          3.97   
4             NaN       Hardcover  9.780000e+12       334          4.47   
...           ...             ...           ...        ...          ...   
1136          NaN       Hardcover  9.780000e+12       304          3.33   
1137          NaN       Hardcover  9.780000e+12       417          3.69   
1138          NaN  Kindle Edition           NaN       320          3.40   
1139          NaN       Hardcover  9.780000e+12       419          3.86   
1140          NaN       Hardcover  9.780000e+12       304          3.64   

      book_rating_count  book_review_count  \
0                146505              17438   
1                 61357               9369   
2                 62897              11948   
3                175678              20645   
4                128337              16424   
...                 ...                ...   
1136                359                 88   
1137               1255                259   
1138               1300                272   
1139                907                205   
1140               1322                395   

                              book_title  \
0                        The Great Alone   
1                                  Circe   
2                       The Cruel Prince   
3                The Woman in the Window   
4                               Educated   
...                                  ...   
1136                        Bound to You   
1137                     If You Leave Me   
1138  This Love Story Will Self-Destruct   
1139                 The Queen of Sorrow   
1140         My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                       Fantasy|Young Adult|Fairies|Fae   
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4             Nonfiction|Autobiography|Memoir|Biography   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url  
0     https://images.gr-assets.com/books/1501852423l...  
1     https://images.gr-assets.com/books/1508879575l...  
2     https://images.gr-assets.com/books/1493047153l...  
3     https://images.gr-assets.com/books/1528225499l...  
4     https://images.gr-assets.com/books/1506026635l...  
...                                          

In [238]:
book_data['book_pages'].astype(int)

0       435
1       393
2       370
3       449
4       334
       ... 
1136    304
1137    417
1138    320
1139    419
1140    304
Name: book_pages, Length: 1071, dtype: int64

In [239]:
book_data.dtypes

book_authors          object
book_desc             object
book_edition          object
book_format           object
book_isbn            float64
book_pages            object
book_rating          float64
book_rating_count      int64
book_review_count      int64
book_title            object
genres                object
image_url             object
dtype: object

### Q6: What is the average number of pages? (5 points)

In [244]:
book_page=book_data['book_pages'].astype(float)
book_page

0       435.0
1       393.0
2       370.0
3       449.0
4       334.0
        ...  
1136    304.0
1137    417.0
1138    320.0
1139    419.0
1140    304.0
Name: book_pages, Length: 1071, dtype: float64

In [245]:
book_page.mean()

327.9000933706816

### Q7: What is the median number of pages? (5 points)

In [246]:
book_page.median()

328.0

### Q8: Does having more than 1 author result in a longer book on average? What is the average number of pages for books written by a single author? What is it for when there are at least two authors? (We will do more careful analysis for these types of questions later. For now, we just want you to practice using some DataFrame functionalities). (5 points)!

In [ ]:
# 先暂停
author_num

In [248]:
for author_num in book_data["book_pages"]:
    if author_num > 1:
        print(book_data["book_pages"].mean())
        print(book_data["book_pages"].median())

TypeError: '>' not supported between instances of 'str' and 'int'

## Ratings: (30 points)

### Q9: How many books have at least a rating of 4.5? (2.5 points)

### Q10: How about at most a rating of 4.25? (2.5 points)

### Q11: Discretize the ratings. (1,2,3,4,5). (15 points)

### Q12: For each of the discretized ratings (1,2,3,4,5), what is the average number of ratings? (5 points)

### Q13: How about the average number of authors? (5 points)

## Genres: (35 points)

### Q14: Create a new DataFrame, exploding the rows with multiple genres such that it is one row per genre/book. (15 points)

### Q15: What is the average rating of books from different genres? What is the median? (10 points)

### Q16: What is the average length of books from different genres? What is the median? (10 points)